frontend/app/src/MetricsManager.ts

```diff
@@ -0,0 +1,374 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import pick from "lodash/pick"
+import { v4 as uuidv4 } from "uuid"
+import { IGuestToHostMessage } from "lib/src/hostComm/types"
+
+import { initializeSegment } from "@streamlit/app/src/vendor/Segment"
+import {
+  DeployedAppMetadata,
+  getCookie,
+  IS_DEV_ENV,
+  localStorageAvailable,
+  logAlways,
+  logError,
+  MetricsEvent,
+  SessionInfo,
+  setCookie,
+} from "@streamlit/lib"
+
+// Default metrics config fetched when none provided by host config endpoint
+export const DEFAULT_METRICS_CONFIG = "https://data.streamlit.io/metrics.json"
+
+/**
+ * The analytics is the Segment.io object. It is initialized in Segment.ts
+ * It is loaded with global scope (window.analytics) to integrate with the segment.io api
+ * @global
+ * */
+declare const analytics: any
+
+type Event = [string, Record<string, unknown>]
```

Comment:
Since the PR adds a more strictly defined Metrics object (`MetricsEvent.proto`), it could be used to enforce stricter typing within the Metrics Manager. E.g. 

```typescript
import {
   ...
  IMetricsEvent
} from "@streamlit/lib"

...

type EventName = "pageProfile" | "menuClick" | "updateReport" | "viewReport"
type Event = [EventName, Partial<IMetricsEvent>]
```

And it could be used to type the event name & data in all the methods in this class:

```
private send(evName: EventName, evData: Partial<IMetricsEvent> = {}): void {
...
```

This would help ensure that properties are only added to the events if they are also added to the `MetricsEvent.proto`.